polling/poller.py

```diff
@@ -0,0 +1,106 @@
+    #       Remember that end_time could be None
+    #       sites could have one sites, many sites, or all of your sites
+    print(
+        "Dispatch with ID {} starting at {} and ending at {} for sites {} created".format(
+            dispatch_info["id"],
+            dispatch_info["start_time"],
+            dispatch_info.get("end_time", "not specified"),
+            dispatch_info["sites"],
+        )
+    )
+
+
+def update_dispatch(dispatch_info: Dict):
+    # TODO: Fill in with your company-specific logic
+    #       start_time and end_time can change to be earlier or later
+    #       Remember that this could get called twice if the service is restarted or the request otherwise fails
+    print(
+        "Dispatch with ID {} starting at {} and ending at {} for sites {} updated".format(
+            dispatch_info["id"],
+            dispatch_info["start_time"],
+            dispatch_info.get("end_time", "not specified"),
+            dispatch_info["sites"],
+        )
+    )
+
+
+def cancel_dispatch(dispatch_info: Dict):
+    # TODO: Fill in with your company-specific logic to cancel this dispatch, likely by ID
+    #       Remember that this could get called twice if the service is restarted
+    print("Dispatch with ID {} cancelled".format(dispatch_info["id"]))
+
+
+if __name__ == "__main__":
+    while True:
+        start_time = datetime.datetime.now(datetime.timezone.utc)
+
+        url = urllib.parse.urljoin(VOLTUS_API_URL, "/2022-04-15/dispatches")
+        resp = s.get(url)
+        resp.raise_for_status()  # raise an exception if request fails
+        dispatches = resp.json()
+        for dispatch_info in dispatches["dispatches"]:
+            # convert start/end time to proper datetime
+            dispatch_info["start_time"] = date_parser.parse(dispatch_info["start_time"])
+            if "end_time" in dispatch_info and dispatch_info["end_time"] is not None:
+                dispatch_info["end_time"] = date_parser.parse(dispatch_info["end_time"])
+            else:
+                dispatch_info["end_time"] = None
+
+            dispatch_id = dispatch_info["id"]
+            now = datetime.datetime.now(datetime.timezone.utc)
+
+            # if we've already cancelled this dispatch, skip it
+            if dispatch_id in cancelled_dispatches:
+                continue
+
+            # send a cancelled message if the dispatch is over or not authorized
+            # regardless of whether we know about it or not (to protect against
+            # this service restarting or similar), as long as it has not been
+            # previously marked as cancelled
+            if (
+                dispatch_info.get("end_time") and dispatch_info["end_time"] < now
+            ) or not dispatch_info["authorized"]:
+                cancel_dispatch(dispatch_info)
+                del managed_dispatches[dispatch_id]
+                cancelled_dispatches.add(dispatch_id)
```

Comment:
Nit, very optional: put adding to cancelled before removing from managed for resiliency

Reply:
I don't think this matters all that much, though now that you mention it the `del` statement would throw an exception if we haven't seen this dispatch_id before. I'll fix that.

Reply:
Oh - good catch